### DISCUSSION TITLE
How can I get all the messages first, then get only new ones

### DISCUSSION THREAD

**Discussion by Primus-max** [egramtel/tdsharp#166] [Q&A]
Now I have a curved code for receiving messages, the first time I use 
var result = await _client. ExecuteAsync(new TdApi.GetChatHistory
                {
                    ChatId = chatId,
                    Limit = 100,
                    Offset = -1,
                    FromMessageId = lastMessageId
                }); 
Then I use the rest to get the rest
 var result = await _client.ExecuteAsync(new TdApi.SearchChatMessages
                    {
                        ChatId = chatId,
                        Limit = 100,
                        FromMessageId = lastMessageId,
                    });
But even so, I don't get everything I need, and often the receipt is interrupted and does not continue.
I tried only to get var result = await _client. ExecuteAsync(new TdApi.GetChatHistory
                {
                    ChatId = chatId,
                    Limit = 100,
                    Offset = -1,
                    FromMessageId = lastMessageId
                }); 
But with any settings I get only 1 message, sometimes it is more.

**Comment by ForNeVeR** [comment]
Could you please clarify what's the particular problem? Do you have issues with receiving the messages (if so, then which ones)? Or with something being "interrupted"?

**Comment by Primus-max** [comment]
Once I need to get all the messages. Once a day I need to receive messages from the last messageId. I do this 
var result = await _client. ExecuteAsync(new SearchChatMessages
                {
                    ChatId = chatId,
                    Limit = 100,
                    FromMessageId = lastMessageId,
                });
 lastMessageId - here I pass the last id if it is, otherwise 0. But as I understand it, messageId is a unique number and in theory it cannot be = 0. Collecting messages starts with the first message that came last, but I need to collect from the very first.

**Reply by ForNeVeR** [reply]
According [to the documentation](https://core.telegram.org/tdlib/docs/classtd_1_1td__api_1_1search_chat_messages.html), you are doing that correctly. If it doesn't work for you, then please [open an issue on the upstream repository](https://github.com/tdlib/td).

**Reply by Primus-max** [reply]
The problem is that at some point the collection begins again or continues, but the folder with the downloaded photos is deleted and the download begins again. I just need a photo. 6000 downloaded, I was distracted, when I looked at the photo was already 3000.
And the memory is badly clogged.

**Reply by Primus-max** [reply]
And a very large memory leak in the var downloadedFile=await method _client. DownloadFileAsync(largestSize.Photo.Id, 1, 0, 0, true); 
The last parameter = true because false does not get localImagePath = downloadedFile.Local.Path; 
this value is simply left blank

**Reply by ForNeVeR** [reply]
Could you please clarify what you refer to as a "memory leak"? Have you profiled the code?

**Comment by Primus-max** [comment]
``` cSharp
if (message.Content is MessagePhoto messagePhoto)
                    {
                        try
                        {                            
                            var largestSize = messagePhoto.Photo.Sizes.OrderByDescending(s => s.Width).FirstOrDefault();

                            if (largestSize != null)
                            {                              
                                var downloadedFile = await _client.DownloadFileAsync(largestSize.Photo.Id, 1, 0, 0, true);
                                Thread.Sleep(300);

                                if (downloadedFile.Local.IsDownloadingCompleted)
                                {
                                    localImagePath = downloadedFile.Local.Path;
                                }
                            }
                        }
                        catch (Exception) {  }
                    }
 ```
After 100 uploaded photos, the consumption of RAM increases literally with each uploaded picture. The program starts with 60mb and then increases to 500.

**Reply by ForNeVeR** [reply]
So, what does the profiler say?

**Comment by Primus-max** [comment]
The debugger (if you ask about it) indicates a large memory consumption on TdLib.TdApi. 
Maybe you will see the full method? I know that even with Newtonsoft is the wrong approach, 
but this is not a significant part of the leak. Another problem is that all messages from the group are not fully collected. 
``` cSharp
public async Task ExtractContentFromMessagesAsync(Message message, long chatId)
            {                
                string dataDir = System.IO.Path.Combine(Directory.GetCurrentDirectory(), "content");
                Directory.CreateDirectory(dataDir);
                
                string contentFileName = $"{chatId}.json";
                string contentFilePath = System.IO.Path.Combine(dataDir, contentFileName);

                List<Content>? _contents = null;
                if (File.Exists(contentFilePath))
                {                    
                    try
                    {
                        string jsonContent = await File.ReadAllTextAsync(contentFilePath);
                        _contents = JsonConvert.DeserializeObject<List<Content>>(jsonContent);
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"Fault: {ex.Message}");
                    }
                }
                
                if (_contents == null)
                {
                    _contents = new List<Content>();
                }
              
                bool contentExists = _contents.Any(c => c.MessageId == message.Id);
                if (!contentExists)
                {
                    string? messageText = null;
                    string localImagePath = string.Empty;
                  
                    if (message.Content is MessagePhoto messagePhotoContent)
                    {
                        var captionText = messagePhotoContent.Caption?.Text;
                        if (!string.IsNullOrEmpty(captionText))
                        {
                            messageText = captionText;
                        }
                    }
                    
                    if (message.Content is MessagePhoto messagePhoto)
                    {
                        try
                        {                            
                            var largestSize = messagePhoto.Photo.Sizes.OrderByDescending(s => s.Width).FirstOrDefault();

                            if (largestSize != null)
                            {                                
                                var downloadedFile = await _client.DownloadFileAsync(largestSize.Photo.Id, 1, 0, 0, true);
                                Thread.Sleep(300);

                                if (downloadedFile.Local.IsDownloadingCompleted)
                                {
                                    localImagePath = downloadedFile.Local.Path;
                                }
                            }
                        }
                        catch (Exception ex)
                        {
                            Console.WriteLine($"Fault: {ex.Message}");
                        }
                    }
                    
                    var newContent = new Content()
                    {
                        MessageId = message.Id,
                        Text = messageText,
                        ImagePaths = new List<string>(),
                        ImageViewed = false,
                        MediaAlbumId = message.MediaAlbumId
                    };

                    if (message.MediaAlbumId == 0)
                    {
                        newContent.ImagePaths.Add(localImagePath);
                    }
                    else
                    {
                        var albumContent = _contents.FirstOrDefault(c => c.MediaAlbumId == message.MediaAlbumId);
                        if (albumContent != null)
                        {
                            albumContent?.ImagePaths?.Add(localImagePath);
                        }
                    }

                    _contents.Add(newContent);
                   
                    string jsonContentString = JsonConvert.SerializeObject(_contents, Formatting.Indented);
                    
                    await File.WriteAllTextAsync(contentFilePath, jsonContentString);
                }
            }
 ```

**Reply by ForNeVeR** [reply]
I am not asking about the debugger. I am asking about the profiler. What does the memory profiler show?

**Reply by Primus-max** [reply]
I didn't understand how to upload a photo
https://ibb.co/82SQ88K
https://ibb.co/4ZvLP5v
https://ibb.co/129VLzd
https://ibb.co/c3sJyp9

**Reply by ForNeVeR** [reply]
Ok, this shows there's about 300 KiB of `TextEntityTypeBold`. Definitely not 500 MiB of anything.

I don't think this is actionable for tdsharp in this form, unfortunately. You'll have to investigate your memory profile yourself.
